scripts/cli_regression_tests.py

```diff
@@ -0,0 +1,187 @@
+
+    def test_streamlit_version(self):
+        assert (
+            STREAMLIT_RELEASE_VERSION != None
+        ), "You must set the $STREAMLIT_RELEASE_VERSION env variable"
+        assert STREAMLIT_RELEASE_VERSION in self.run_command("streamlit version")
+
+    def test_streamlit_activate(self):
+        process = subprocess.Popen(
+            "streamlit activate", stdin=subprocess.PIPE, shell=True
+        )
+        process.stdin.write(b"regressiontest@streamlit.io\n")
+        process.stdin.flush()
+        process.communicate()
+
+        with open(CREDENTIALS_FILE_PATH) as f:
+            assert "regressiontest@streamlit.io" in f.read()
+
+    # ---------------------------------------------------------------------------------------------
+    def run_single_proc(self, command, wait_in_seconds=2):
+        proc = subprocess.Popen(
+            command,
+            shell=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+
+        # Sleep to allow commands to run
+        time.sleep(wait_in_seconds)
+
+        # Kill both processes, as output is accessible afterwards
+        proc.terminate()
+
+        try:
+            # Retrieve and return outputs
+            out_one, _ = proc.communicate(timeout=1)
+
+            return out_one.decode("utf-8")
+        except subprocess.TimeoutExpired:
+            assert False, "Subprocess timed out"
+
+    def run_double_proc(self, command_one, command_two, wait_in_seconds=2):
+        proc_one = subprocess.Popen(
+            command_one,
+            shell=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+
+        # Quick sleep to ensure that proc_one gets started first
+        time.sleep(0.1)
+
+        proc_two = subprocess.Popen(
+            command_two,
+            shell=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+
+        # Sleep to allow commands to run
+        time.sleep(wait_in_seconds)
+
+        # Kill both processes, as output is accessible afterwards
+        proc_one.terminate()
+        proc_two.terminate()
+
+        try:
+            # Retrieve and return outputs
+            out_one, _ = proc_one.communicate(timeout=1)
+            out_two, _ = proc_two.communicate(timeout=1)
+
+            return out_one.decode("utf-8"), out_two.decode("utf-8")
+        except subprocess.TimeoutExpired:
+            assert False, "Subprocess timed out"
+
+    def test_port_reassigned(self):
+        """When starting a new Streamlit session, it will run on port 8501 by default. If 8501 is
+        not available, it will use the next available port.
+        """
+
+        out_one, out_two = self.run_double_proc(
+            f"streamlit run --server.headless=true {REPO_ROOT}/examples/file_uploader.py",
+            f"streamlit run --server.headless=true {REPO_ROOT}/examples/file_uploader.py",
+        )
+
+        assert ":8501" in out_one, out_one
```

Comment:
I am unfamiliar with why we'd check that the string is in `out_one, out_one` rather than just `out_one`. Could you add a comment explaining it?

Reply:
Ah, yes. I can document that. The last argument gets logged to the console, which is super helpful if you're debugging a failure and need to see additional context or the data that was used. Here's an example of calling `assert` in an interpreter:

```
>>> assert False, "Not True!"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Not True!
```